### PR TITLE
Extract_geographical_watershed_properties.ipynb: fix run failure in new Jupyter env

### LINKED ISSUE #388: Geospatial toolbox rasterio deprecated code?
There is a notebook (Extract geographical watershed properties) that has a block of code as follows:
____
```
# Note that geojson, and rasterio/rioxarray need to be installed for this to work.
# In the next birdy release, the tiff grid will be automatically converted to a DataArray.
# Here we need to manually convert it.
features, statistics, grid = stats_resp.get(asobj=True)

with rasterio.io.MemoryFile(grid[0]) as memfile:
    with memfile.open() as dataset:
        grid = rio.open_rasterio(dataset)

grid.plot()
```
____

I am trying to write some documentation and have no idea what to do with this. Is this already integrated? Should users know about this? Thanks!

### PR DESCRIPTION
**Pull request by tlvu**
Fixes https://github.com/Ouranosinc/raven/issues/392 using hints from https://github.com/Ouranosinc/raven/issues/388. Closes #388.

Note I am not sure what I am doing as I do not understand the code.  I simply try to get around the errors and produce similar graph.  The old graph had `spatial_ref` in the title, the new one don't.  Wonder if it's something critical.

Clarification from Trevor: 

 ==========
Putting it simply, the object returned (`xarray.DataArray`) is different from what the following cell needed to handle for (`rasterio._io.RasterReader` → `numpy.ndarray`):
```
23:00:58  Input:
23:00:58  # Note that geojson, and rasterio/rioxarray need to be installed for this to work.
23:00:58  # In the next birdy release, the tiff grid will be automatically converted to a DataArray.
23:00:58  # Here we need to manually convert it.
23:00:58  features, statistics, grid = stats_resp.get(asobj=True)
23:00:58
23:00:58  with rasterio.io.MemoryFile(grid[0]) as memfile:
23:00:58      with memfile.open() as dataset:
23:00:58          grid = rio.open_rasterio(dataset)
23:00:58
23:00:58  grid.plot()
```
The `spatial_ref`title is something that `rasterio` adds to the plot, and it isn't at all needed. The notebook should work perfectly so long as the client running it has `rioxarray`available in their ipython kernel.

 ==========

With the new Jupypter env for latest RavenPy `0.5.2` from PR https://github.com/Ouranosinc/PAVICS-e2e-workflow-tests/pull/75 `Extract_geographical_watershed_properties.ipynb` fails with the following error below.

```
23:00:58  _ raven-master/docs/source/notebooks/Extract_geographical_watershed_properties.ipynb::Cell 7 _
23:00:58  Notebook cell execution failed
23:00:58  Cell 7: Cell execution caused an exception
23:00:58
23:00:58  Input:
23:00:58  # Note that geojson, and rasterio/rioxarray need to be installed for this to work.
23:00:58  # In the next birdy release, the tiff grid will be automatically converted to a DataArray.
23:00:58  # Here we need to manually convert it.
23:00:58  features, statistics, grid = stats_resp.get(asobj=True)
23:00:58
23:00:58  with rasterio.io.MemoryFile(grid[0]) as memfile:
23:00:58      with memfile.open() as dataset:
23:00:58          grid = rio.open_rasterio(dataset)
23:00:58
23:00:58  grid.plot()
23:00:58
23:00:58  Traceback:
23:00:58
23:00:58  ---------------------------------------------------------------------------
23:00:58  ValueError                                Traceback (most recent call last)
23:00:58  <ipython-input-8-8b167a3ecdc9> in <module>
23:00:58        4 features, statistics, grid = stats_resp.get(asobj=True)
23:00:58        5
23:00:58  ----> 6 with rasterio.io.MemoryFile(grid[0]) as memfile:
23:00:58        7     with memfile.open() as dataset:
23:00:58        8         grid = rio.open_rasterio(dataset)
23:00:58
23:00:58  /opt/conda/envs/birdy/lib/python3.7/site-packages/rasterio/io.py in __init__(self, file_or_bytes, dirname, filename, ext)
23:00:58      103         """
23:00:58      104         super(MemoryFile, self).__init__(
23:00:58  --> 105             file_or_bytes=file_or_bytes, dirname=dirname, filename=filename, ext=ext)
23:00:58      106
23:00:58      107     @ensure_env
23:00:58
23:00:58  rasterio/_io.pyx in rasterio._io.MemoryFileBase.__init__()
23:00:58
23:00:58  /opt/conda/envs/birdy/lib/python3.7/site-packages/xarray/core/common.py in __bool__(self)
23:00:58      127
23:00:58      128     def __bool__(self: Any) -> bool:
23:00:58  --> 129         return bool(self.values)
23:00:58      130
23:00:58      131     def __float__(self: Any) -> float:
23:00:58
23:00:58  ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
23:00:58
23:00:58  _ raven-master/docs/source/notebooks/Extract_geographical_watershed_properties.ipynb::Cell 10 _
23:00:58  Notebook cell execution failed
23:00:58  Cell 10: Cell execution caused an exception
23:00:58
23:00:58  Input:
23:00:58  unique, counts = numpy.unique(grid.sel(band=1), return_counts=True)
23:00:58  display(unique, counts)
23:00:58
23:00:58  # Pixels values at '127' are NaN and can be ignored.
23:00:58  plt.bar(unique[:-1], counts[:-1])
23:00:58  plt.show()
23:00:58
23:00:58  grid.where(grid != 127).sel(band=1).plot.imshow(cmap="tab20")
23:00:58  plt.show()
23:00:58
23:00:58  Traceback:
23:00:58
23:00:58  ---------------------------------------------------------------------------
23:00:58  AttributeError                            Traceback (most recent call last)
23:00:58  <ipython-input-11-a26c52126e9c> in <module>
23:00:58  ----> 1 unique, counts = numpy.unique(grid.sel(band=1), return_counts=True)
23:00:58        2 display(unique, counts)
23:00:58        3
23:00:58        4 # Pixels values at '127' are NaN and can be ignored.
23:00:58        5 plt.bar(unique[:-1], counts[:-1])
23:00:58
23:00:58  AttributeError: 'list' object has no attribute 'sel'
23:00:58
23:00:58  _ raven-master/docs/source/notebooks/Extract_geographical_watershed_properties.ipynb::Cell 11 _
23:00:58  Notebook cell execution failed
23:00:58  Cell 11: Cell execution caused an exception
23:00:58
23:00:58  Input:
23:00:58  import cartopy.crs as ccrs
23:00:58
23:00:58  # Set a CRS transformation:
23:00:58  crs = ccrs.LambertConformal(
23:00:58      central_latitude=49, central_longitude=-95, standard_parallels=(49, 77)
23:00:58  )
23:00:58
23:00:58  ax = plt.subplot(projection=crs)
23:00:58  grid.name = "Land Use Categories"
23:00:58  grid.where(grid != 127).sel(band=1).plot.imshow(ax=ax, transform=crs, cmap="tab20")
23:00:58  plt.show()
23:00:58
23:00:58  Traceback:
23:00:58
23:00:58  ---------------------------------------------------------------------------
23:00:58  AttributeError                            Traceback (most recent call last)
23:00:58  <ipython-input-12-f9fdb6f0e562> in <module>
23:00:58        7
23:00:58        8 ax = plt.subplot(projection=crs)
23:00:58  ----> 9 grid.name = "Land Use Categories"
23:00:58       10 grid.where(grid != 127).sel(band=1).plot.imshow(ax=ax, transform=crs, cmap="tab20")
23:00:58       11 plt.show()
23:00:58
23:00:58  AttributeError: 'list' object has no attribute 'name'
```

### CODE REVIEW

**Comment by Zeitsperre** [comment]
I'll take a look at this this afternoon!

File: docs/source/notebooks/Extract_geographical_watershed_properties.ipynb [hunk 1]
```diff
@@ -1044,7 +1043,7 @@
     {
      "data": {
       "text/plain": [
-       "'https://pavics.ouranos.ca/wpsoutputs/1835706e-ca30-11eb-934d-0242ac120017/input.json'"
+       "'http://localhost:9099/outputs/59d9f8a2-cb00-11eb-94d3-28d24417f566/input.json'"
```
Comment (tlvu) [inline docs/source/notebooks/Extract_geographical_watershed_properties.ipynb:1046]: Can you re-run using pavics and not localhost?

File: docs/source/notebooks/Extract_geographical_watershed_properties.ipynb [hunk 2]
```diff
@@ -324,13 +322,14 @@
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "Downloading to /tmp/tmpq5dx3ekj/subset_1.tiff.\n"
+      "Metalink content-type detected.\n",
```
Comment (tlvu) [inline docs/source/notebooks/Extract_geographical_watershed_properties.ipynb:325]: Are you on the latest birdy?  Funny that the output has the additional "Metalink content-type detected"?

File: docs/source/notebooks/Extract_geographical_watershed_properties.ipynb [hunk 3]
```diff
@@ -1077,7 +1076,7 @@
    ],
    "source": [
     "# We can also access the files directly via their URLs:\n",
-    "properties, dem = terrain_resp.get(asobj=False)\n",
+    "properties, dem = terrain_resp.get()\n",
```
Comment (tlvu) [inline docs/source/notebooks/Extract_geographical_watershed_properties.ipynb:1079]: nice catch, asjob=False is like it's not there.

**Review by tlvu — COMMENTED** [review]
Can you re-run and hit the production pavics server instead.

Alternatively, use `make docs/source/notebooks/Extract_geographical_watershed_properties.ipynb.refresh` to re-run against localhost but all `localhost` will be replaced for you automatically.

Otherwise, that's for all the explanation !

File: docs/source/notebooks/Extract_geographical_watershed_properties.ipynb (hunk 2 above)
Comment (Zeitsperre) [reply docs/source/notebooks/Extract_geographical_watershed_properties.ipynb:325]: I setup a fresh installation from the `environment.yml`. It must be the newest version.

**Comment by Zeitsperre** [comment]
> Alternatively, use `make docs/source/notebooks/Extract_geographical_watershed_properties.ipynb.refresh` to re-run against localhost but all `localhost` will be replaced for you automatically.

I think the method of file path parsing needs some tweaking. Failing with:
```
FileNotFoundError: [Errno 2] No such file or directory: 'docs/source/notebooks/docs/source/notebooks/Extract_geographical_watershed_properties.ipynb'
```

Will try re-running manually for now.

**Comment by tlvu** [comment]
> > Alternatively, use `make docs/source/notebooks/Extract_geographical_watershed_properties.ipynb.refresh` to re-run against localhost but all `localhost` will be replaced for you automatically.
> 
> I think the method of file path parsing needs some tweaking. Failing with:
> 
> ```
> FileNotFoundError: [Errno 2] No such file or directory: 'docs/source/notebooks/docs/source/notebooks/Extract_geographical_watershed_properties.ipynb'
> ```
> 

@Zeitsperre How did you get this error?  What was your `$PWD` when you launch `make`?

> Will try re-running manually for now.

It works, passing Jenkins for all Raven notebooks http://jenkins.ouranos.ca/job/PAVICS-e2e-workflow-tests/job/new-docker-build-for-ravenpy/18/console, thanks !

@richardarsenault given all Raven notebooks working now, if we deploy a new Jupyter env, is it okay for your summer school?

**Comment by richardarsenault** [comment]
@tlvu Yes it's all good for the summer school, thanks!

File: docs/source/notebooks/Extract_geographical_watershed_properties.ipynb [hunk 4]
```diff
@@ -316,20 +365,26 @@
...
+     "iopub.status.idle": "2021-06-14T14:05:42.497511Z",
+     "shell.execute_reply": "2021-06-14T14:05:42.496989Z"
+    }
+   },
    "outputs": [
     {
      "name": "stdout",
      "output_type": "stream",
      "text": [
-      "Metalink content-type detected.\n",
```
Comment (tlvu) [inline docs/source/notebooks/Extract_geographical_watershed_properties.ipynb:327]: Oh cool, this latest manual re-run removed "Metalink content-type detected.".

Comment (Zeitsperre) [reply docs/source/notebooks/Extract_geographical_watershed_properties.ipynb:327]: Yeah, I wonder why that is. I didn't do anything differently.

**Comment by Zeitsperre** [comment]
> > FileNotFoundError: [Errno 2] No such file or directory: 'docs/source/notebooks/docs/source/notebooks/Extract_geographical_watershed_properties.ipynb'
> 
> @Zeitsperre How did you get this error? What was your `$PWD` when you launch `make`?

`/my_user/git/raven/`. Here's the full stack:
```
Refreshing notebook docs/source/notebooks/Extract_geographical_watershed_properties.ipynb
make[1]: Entering directory '/home/my_user/git/raven'
bash -c 'WPS_URL="http://0.0.0.0:9099" FINCH_WPS_URL="https://pavics.ouranos.ca/twitcher/ows/proxy/finch/wps" FLYINGPIGEON_WPS_URL="https://pavics.ouranos.ca/twitcher/ows/proxy/flyingpigeon/wps" jupyter nbconvert --to notebook --execute --ExecutePreprocessor.timeout=240 --output "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb" "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb"; sed -i "s@http://0.0.0.0:9099/outputs/@https://pavics.ouranos.ca/wpsoutputs/@g" "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb"'
[NbConvertApp] Converting notebook docs/source/notebooks/Extract_geographical_watershed_properties.ipynb to notebook
[NbConvertApp] Executing notebook with kernel: python3
[NbConvertApp] Writing 144740 bytes to docs/source/notebooks/docs/source/notebooks/Extract_geographical_watershed_properties.ipynb
Traceback (most recent call last):
  File "/home/my_user/miniconda3/envs/raven-dev1/bin/jupyter-nbconvert", line 11, in <module>
    sys.exit(main())
  File "/home/my_user/miniconda3/envs/raven-dev1/lib/python3.8/site-packages/jupyter_core/application.py", line 254, in launch_instance
    return super(JupyterApp, cls).launch_instance(argv=argv, **kwargs)
  File "/home/my_user/miniconda3/envs/raven-dev1/lib/python3.8/site-packages/traitlets/config/application.py", line 845, in launch_instance
    app.start()
  File "/home/my_user/miniconda3/envs/raven-dev1/lib/python3.8/site-packages/nbconvert/nbconvertapp.py", line 350, in start
    self.convert_notebooks()
  File "/home/my_user/miniconda3/envs/raven-dev1/lib/python3.8/site-packages/nbconvert/nbconvertapp.py", line 524, in convert_notebooks
    self.convert_single_notebook(notebook_filename)
  File "/home/my_user/miniconda3/envs/raven-dev1/lib/python3.8/site-packages/nbconvert/nbconvertapp.py", line 490, in convert_single_notebook
    write_results = self.write_single_notebook(output, resources)
  File "/home/my_user/miniconda3/envs/raven-dev1/lib/python3.8/site-packages/nbconvert/nbconvertapp.py", line 450, in write_single_notebook
    write_results = self.writer.write(
  File "/home/my_user/miniconda3/envs/raven-dev1/lib/python3.8/site-packages/nbconvert/writers/files.py", line 126, in write
    with io.open(dest, 'w', encoding='utf-8') as f:
FileNotFoundError: [Errno 2] No such file or directory: 'docs/source/notebooks/docs/source/notebooks/Extract_geographical_watershed_properties.ipynb'
make[1]: Leaving directory '/home/my_user/git/raven'
```
Could it be that this particular `make` command doesn't work in `zsh`? I can't think of why else it would fail.

**Comment by tlvu** [comment]
> Could it be that this particular `make` command doesn't work in `zsh`? I can't think of why else it would fail.
> 
> EDIT: Tried from `bash`. No difference in output.

This is really odd.  This works for me on my machine.

Can you try this command direct in your shell?
```
bash -c 'WPS_URL="http://0.0.0.0:9099" FINCH_WPS_URL="https://pavics.ouranos.ca/twitcher/ows/proxy/finch/wps" FLYINGPIGEON_WPS_URL="https://pavics.ouranos.ca/twitcher/ows/proxy/flyingpigeon/wps" jupyter nbconvert --to notebook --execute --ExecutePreprocessor.timeout=240 --output "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb" "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb"; sed -i "s@http://0.0.0.0:9099/outputs/@https://pavics.ouranos.ca/wpsoutputs/@g" "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb"'
```

**Comment by Zeitsperre** [comment]
> Can you try this command direct in your shell?
> 
> ```
> bash -c 'WPS_URL="http://0.0.0.0:9099" FINCH_WPS_URL="https://pavics.ouranos.ca/twitcher/ows/proxy/finch/wps" FLYINGPIGEON_WPS_URL="https://pavics.ouranos.ca/twitcher/ows/proxy/flyingpigeon/wps" jupyter nbconvert --to notebook --execute --ExecutePreprocessor.timeout=240 --output "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb" "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb"; sed -i "s@http://0.0.0.0:9099/outputs/@https://pavics.ouranos.ca/wpsoutputs/@g" "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb"'
> ```

https://gist.github.com/Zeitsperre/e430dda569c6e93ff9903d7c323020e3

**Comment by Zeitsperre** [comment]
So it looks like this works with nbconvert 6.0.7. `--output` must only be the name of the file. `--output-dir` is to specify the new location, but by default it will use the same location as source.:
```
WPS_URL="http://0.0.0.0:9099" FINCH_WPS_URL="https://pavics.ouranos.ca/twitcher/ows/proxy/finch/wps" FLYINGPIGEON_WPS_URL="https://pavics.ouranos.ca/twitcher/ows/proxy/flyingpigeon/wps" jupyter nbconvert 
--to notebook --execute --ExecutePreprocessor.timeout=240 --output "Extract_geographical_watershed_properties.ipynb" "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb"; sed -i "s@http://0.0.0.0:9099/outputs/@https://pavics.ouranos.ca/wpsoutputs/@g" "docs/source/notebooks/Extract_geographical_watershed_properties.ipynb" 
```

**Comment by tlvu** [comment]
Merged anyways since the build error not related to the notebook I changed:
```
tests/test_wps_hydrobasins_shape_selection.py .
/home/runner/work/_temp/09d6c889-d3e4-47f8-a805-4f001914160b.sh: line 4:  4210 Segmentation fault      (core dumped) env RAVENPY_TESTDATA_PATH=$GITHUB_WORKSPACE/raven-testdata pytest tests
```

**Pull request merged**